From a8a640d2a929fd6cdb13b5e8687288d2467ea5d4 Mon Sep 17 00:00:00 2001
From: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>
Date: Fri, 10 Jan 2014 17:37:54 +0800
Subject: [PATCH 1/2] virtinst: add support for panic notifier device

Signed-off-by: Chen Hanxiao <chenhanxiao@cn.fujitsu.com>
(cherry picked from commit 8419835b1d68e5bfad1005b00c87d302279a118f)

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1101536

Conflicts:
    virtinst/__init__.py
    virtinst/VirtualPanicDevice.py
---
 virtinst/Guest.py              |  1 +
 virtinst/VirtualDevice.py      |  4 ++-
 virtinst/VirtualPanicDevice.py | 73 ++++++++++++++++++++++++++++++++++++++++++
 virtinst/__init__.py           |  1 +
 4 files changed, 78 insertions(+), 1 deletion(-)
 create mode 100644 virtinst/VirtualPanicDevice.py

diff --git a/virtinst/Guest.py b/virtinst/Guest.py
index bb79f45..b10126e 100644
--- a/virtinst/Guest.py
+++ b/virtinst/Guest.py
@@ -773,6 +773,7 @@ class Guest(XMLBuilderDomain.XMLBuilderDomain):
             "filesystem": virtinst.VirtualFilesystem,
             "smartcard" : virtinst.VirtualSmartCardDevice,
             "redirdev"   : virtinst.VirtualRedirDevice,
+            "panic"     : virtinst.VirtualPanicDevice,
         }
 
         # Hand off all child element parsing to relevant classes
diff --git a/virtinst/VirtualDevice.py b/virtinst/VirtualDevice.py
index 84ce763..48c0471 100644
--- a/virtinst/VirtualDevice.py
+++ b/virtinst/VirtualDevice.py
@@ -43,6 +43,7 @@ class VirtualDevice(XMLBuilderDomain):
     VIRTUAL_DEV_FILESYSTEM      = "filesystem"
     VIRTUAL_DEV_SMARTCARD       = "smartcard"
     VIRTUAL_DEV_REDIRDEV        = "redirdev"
+    VIRTUAL_DEV_PANIC           = "panic"
 
     # Ordering in this list is important: it will be the order the
     # Guest class outputs XML. So changing this may upset the test suite
@@ -61,7 +62,8 @@ class VirtualDevice(XMLBuilderDomain):
                             VIRTUAL_DEV_HOSTDEV,
                             VIRTUAL_DEV_WATCHDOG,
                             VIRTUAL_DEV_SMARTCARD,
-                            VIRTUAL_DEV_REDIRDEV]
+                            VIRTUAL_DEV_REDIRDEV,
+                            VIRTUAL_DEV_PANIC]
 
     # General device type (disk, interface, etc.)
     _virtual_device_type = None
diff --git a/virtinst/VirtualPanicDevice.py b/virtinst/VirtualPanicDevice.py
new file mode 100644
index 0000000..313e926
--- /dev/null
+++ b/virtinst/VirtualPanicDevice.py
@@ -0,0 +1,73 @@
+# coding=utf-8
+#
+# Copyright 2013 Fujitsu Limited.
+# Copyright 2014  Red Hat, Inc.
+# Chen Hanxiao <chenhanxiao at cn.fujitsu.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+# MA 02110-1301 USA.
+
+import VirtualDevice
+
+from virtinst import _gettext as _
+from XMLBuilderDomain import _xml_property
+
+class VirtualPanicDevice(VirtualDevice.VirtualDevice):
+
+    _virtual_device_type = VirtualDevice.VirtualDevice.VIRTUAL_DEV_PANIC
+    ADDRESS_TYPE_ISA = "isa"
+    TYPES = [ADDRESS_TYPE_ISA]
+    IOBASE_DEFAULT = "0x505"
+
+    def __init__(self, conn, parsexml=None, parsexmlnode=None, caps=None):
+        VirtualDevice.VirtualDevice.__init__(self, conn, parsexml,
+                                             parsexmlnode, caps)
+
+        self._type = None
+        self._iobase = None
+
+        if self._is_parse():
+            return
+
+        self._type = self.ADDRESS_TYPE_ISA
+        self._iobase = self.IOBASE_DEFAULT
+
+    def get_iobase(self):
+        return self._iobase
+    def set_iobase(self, new_iobase):
+        self._iobase = new_iobase
+    iobase = _xml_property(get_iobase, set_iobase,
+                          xpath="./address/@iobase")
+
+    def get_type(self):
+        return self._type
+    def set_type(self, new_type):
+        self._type = new_type
+    type = _xml_property(get_type, set_type,
+                          xpath="./address/@type")
+
+    @staticmethod
+    def get_pretty_type(panic_type):
+        if panic_type == VirtualPanicDevice.ADDRESS_TYPE_ISA:
+            return _("ISA")
+        return panic_type
+
+    def _get_xml_config(self):
+        xml = "    <panic>\n"
+        xml += "      <address type='%s' iobase='%s'/>\n" % (self._type,
+                                                            self._iobase)
+        xml += "    </panic>\n"
+
+        return xml
diff --git a/virtinst/__init__.py b/virtinst/__init__.py
index efa4d0d..93e6a4e 100644
--- a/virtinst/__init__.py
+++ b/virtinst/__init__.py
@@ -51,6 +51,7 @@ from VirtualWatchdog import VirtualWatchdog
 from VirtualFilesystem import VirtualFilesystem
 from VirtualSmartCardDevice import VirtualSmartCardDevice
 from VirtualRedirDevice import VirtualRedirDevice
+from VirtualPanicDevice import VirtualPanicDevice
 from FullVirtGuest import FullVirtGuest
 from ParaVirtGuest import ParaVirtGuest
 from DistroInstaller import DistroInstaller
-- 
1.9.3

